File: bindings/php/src/FaizDB.php
<?php

namespace FaizDB;

/**
 * 🔥 FaizDB Official PHP SDK — The AI-Native NoSQL Database Client.
 */
class FaizDB
{
    private string $endpoint;

    public function __construct(string $endpoint = 'http://localhost:27018')
    {
        $this->endpoint = rtrim($endpoint, '/');
    }

    public function collection(string $name): FaizCollection
    {
        return new FaizCollection($this, $name);
    }

    public function query(string $queryString): mixed
    {
        $res = $this->request('/v1/query', 'POST', ['query' => $queryString]);
        if (!($res['success'] ?? false)) {
            throw new \RuntimeException($res['error'] ?? 'Query execution failed');
        }
        $data = $res['data'] ?? null;
        if (is_array($data)) {
            foreach (['Documents', 'Count', 'Inserted', 'Updated', 'Deleted', 'Success'] as $key) {
                if (array_key_exists($key, $data)) {
                    return $data[$key];
                }
            }
        }
        return $data;
    }

    public function request(string $path, string $method = 'GET', ?array $body = null): array
    {
        $url = $this->endpoint . $path;
        $options = [
            'http' => [
                'header'  => "Content-Type: application/json\r\nAccept: application/json\r\n",
                'method'  => $method,
                'content' => $body ? json_encode($body) : null,
                'ignore_errors' => true,
            ]
        ];
        $context  = stream_context_create($options);
        $result = @file_get_contents($url, false, $context);
        if ($result === false) {
            throw new \RuntimeException("Failed to connect to FaizDB at $url");
        }
        return json_decode($result, true) ?: [];
    }
}

class FaizCollection
{
    private FaizDB $client;
    public string $name;

    public function __construct(FaizDB $client, string $name)
    {
        $this->client = $client;
        $this->name = $name;
    }

    public function insert(array $document): string
    {
        $res = $this->client->request("/v1/collections/{$this->name}/insert", 'POST', $document);
        if (!($res['success'] ?? false)) {
            throw new \RuntimeException($res['error'] ?? 'Insert failed');
        }
        return $res['data']['id'];
    }

    public function find(?array $filter = null): mixed
    {
        $filterStr = $filter ? json_encode($filter) : '{}';
        return $this->client->query("db.{$this->name}.find($filterStr)");
    }

    public function vectorSearch(array $vector, int $topK = 10): mixed
    {
        $vecJson = json_encode($vector);
        return $this->client->query("FIND {$this->name} VECTOR NEAR $vecJson TOP $topK");
    }
}

File: bindings/go/faizdb.go
package faizdb

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

// Client is the FaizDB client
type Client struct {
	endpoint string
	token    string
	http     *http.Client
}

// NewClient creates a new FaizDB client connection
func NewClient(endpoint string, token ...string) *Client {
	if endpoint == "" {
		endpoint = "http://localhost:27018"
	}
	t := ""
	if len(token) > 0 {
		t = token[0]
	}
	return &Client{
		endpoint: strings.TrimRight(endpoint, "/"),
		token:    t,
		http:     &http.Client{Timeout: 30 * time.Second},
	}
}

// Login authenticates with master or user credentials
func (c *Client) Login(username, password string) (string, error) {
	payload, _ := json.Marshal(map[string]string{
		"username": username,
		"password": password,
	})
	resp, err := c.http.Post(c.endpoint+"/v1/auth/login", "application/json", bytes.NewReader(payload))
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	var apiResp struct {
		Success bool `json:"success"`
		Data    struct {
			Token string `json:"token"`
			Role  string `json:"role"`
		} `json:"data"`
		Error string `json:"error"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&apiResp); err != nil {
		return "", err
	}
	if !apiResp.Success {
		return "", fmt.Errorf("login failed: %s", apiResp.Error)
	}
	c.token = apiResp.Data.Token
	return c.token, nil
}

// Collection returns a handle to a collection
func (c *Client) Collection(name string) *Collection {
	return &Collection{
		client: c,
		name:   name,
	}
}

// Query executes any SQL, MongoDB, or FaizQL query string
func (c *Client) Query(queryString string) (interface{}, error) {
	reqBody, _ := json.Marshal(map[string]string{"query": queryString})
	req, err := http.NewRequest("POST", c.endpoint+"/v1/query", bytes.NewReader(reqBody))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	if c.token != "" {
		req.Header.Set("Authorization", "Bearer "+c.token)
	}

	resp, err := c.http.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	var apiResp struct {
		Success bool        `json:"success"`
		Data    interface{} `json:"data"`
		Error   string      `json:"error"`
	}
	if err := json.Unmarshal(body, &apiResp); err != nil {
		return nil, err
	}
	if !apiResp.Success {
		return nil, fmt.Errorf("query error: %s", apiResp.Error)
	}
	return apiResp.Data, nil
}

// Collection handle
type Collection struct {
	client *Client
	name   string
}

// Insert inserts a document
func (col *Collection) Insert(doc map[string]interface{}) (string, error) {
	reqBody, _ := json.Marshal(doc)
	req, err := http.NewRequest(
		"POST",
		fmt.Sprintf("%s/v1/collections/%s/insert", col.client.endpoint, col.name),
		bytes.NewReader(reqBody),
	)
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")
	if col.client.token != "" {
		req.Header.Set("Authorization", "Bearer "+col.client.token)
	}

	resp, err := col.client.http.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	var apiResp struct {
		Success bool              `json:"success"`
		Data    map[string]string `json:"data"`
		Error   string            `json:"error"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&apiResp); err != nil {
		return "", err
	}
	if !apiResp.Success {
		return "", fmt.Errorf("insert failed: %s", apiResp.Error)
	}
	return apiResp.Data["id"], nil
}

// Find executes a find query
func (col *Collection) Find(filter map[string]interface{}) (interface{}, error) {
	filterBytes, _ := json.Marshal(filter)
	query := fmt.Sprintf("db.%s.find(%s)", col.name, string(filterBytes))
	return col.client.Query(query)
}

// VectorSearch runs high-speed HNSW similarity search
func (col *Collection) VectorSearch(vector []float32, topK int) (interface{}, error) {
	vecJson, _ := json.Marshal(vector)
	query := fmt.Sprintf("FIND %s VECTOR NEAR %s TOP %d", col.name, string(vecJson), topK)
	return col.client.Query(query)
}
